Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp

```diff
@@ -0,0 +1,783 @@
+    GRefPtr<RiceAgent> agent;
+
+    Vector<std::unique_ptr<WebKitGstRiceStream>> streams;
+
+    RefPtr<RunLoop> runLoop;
+
+    Atomic<bool> agentIsClosed;
+    GRefPtr<GstPromise> closePromise;
+
+    GstWebRTCICEOnCandidateFunc onCandidate;
+    gpointer onCandidateData;
+    GDestroyNotify onCandidateNotify;
+
+    RefPtr<RiceBackend> iceBackend;
+
+    String stunServer;
+    String turnServer;
+
+    HashSet<URL> turnServers;
+    Vector<GRefPtr<RiceTurnConfig>> turnConfigs;
+} WebKitGstIceAgentPrivate;
+
+typedef struct _WebKitGstIceAgent {
+    GstWebRTCICE parent;
+    WebKitGstIceAgentPrivate* priv;
+} WebKitGstIceAgent;
+
+typedef struct _WebKitGstIceAgentClass {
+    GstWebRTCICEClass parentClass;
+} WebKitGstIceAgentClass;
+
+GST_DEBUG_CATEGORY(webkit_webrtc_rice_debug);
+#define GST_CAT_DEFAULT webkit_webrtc_rice_debug
+
+WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitGstIceAgent, webkit_gst_webrtc_ice_backend, GST_TYPE_WEBRTC_ICE, GST_DEBUG_CATEGORY_INIT(webkit_webrtc_rice_debug, "webkitwebrtcrice", 0, "WebRTC Rice ICE backend"))
+
+using namespace WebCore;
+
+static void webkitGstWebRTCIceAgentSetOnIceCandidate(GstWebRTCICE* ice, GstWebRTCICEOnCandidateFunc callback, gpointer userData, GDestroyNotify notifyCallback)
+{
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(ice);
+    auto priv = backend->priv;
+    if (priv->onCandidateNotify)
+        priv->onCandidateNotify(priv->onCandidateData);
+    priv->onCandidateNotify = notifyCallback;
+    priv->onCandidateData = userData;
+    priv->onCandidate = callback;
+}
+
+static void webkitGstWebRTCIceAgentSetForceRelay(GstWebRTCICE*, gboolean)
+{
+    GST_FIXME("Not implemented yet.");
+}
+
+static void webkitGstWebRTCIceAgentSetStunServer(GstWebRTCICE* ice, const gchar* uri)
+{
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(ice);
+    backend->priv->stunServer = String::fromUTF8(uri);
+    GST_INFO_OBJECT(ice, "Setting STUN server address to %s", uri);
+
+    URL url(backend->priv->stunServer);
+    ASSERT(url.isValid());
+    const auto& host = url.host();
+    auto port = url.port().value_or(3478);
+
+    backend->priv->iceBackend->resolveAddress(host.toString(), [weakAgent = GThreadSafeWeakPtr(backend), port](ExceptionOr<String>&& result) mutable {
+        auto agent = weakAgent.get();
+        if (!agent) [[unlikely]]
+            return;
+        if (result.hasException()) {
+            GST_WARNING_OBJECT(agent.get(), "Unable to configure STUN server on ICE agent: %s", result.exception().message().utf8().data());
+            return;
+        }
+        const auto& iceAgent = agent->priv->agent;
+        if (!iceAgent) [[unlikely]]
+            return;
+
+        auto& host = result.returnValue();
+        auto address = makeString(host, ':', port);
+        GST_DEBUG_OBJECT(agent.get(), "STUN server address resolved to %s", address.ascii().data());
+        GUniquePtr<RiceAddress> stunAddress(rice_address_new_from_string(address.ascii().data()));
```

Comment:
Create the CString in the stack by calling ascii(), otherwise we are duplicating the string twice. Then you just call data on the CString.